packages/core/lib/queues/AbstractQueueService.ts

```diff
@@ -370,4 +375,76 @@ export abstract class AbstractQueueService<
   ): Promise<Either<'retryLater', 'success'>>
 
   public abstract close(): Promise<unknown>
+
+  /**
+   * Offload message payload to an external store if it exceeds the threshold.
+   * Returns a special type that contains a pointer to the offloaded payload or the original payload if it was not offloaded.
+   * Requires message size as only the implementation knows how to calculate it.
+   */
+  protected async offloadMessagePayload(
+    message: MessagePayloadSchemas,
+    messageSize: number,
+  ): Promise<MessagePayloadSchemas | OffloadedPayloadPointerPayload> {
+    if (!this.payloadStoreConfig || messageSize <= this.payloadStoreConfig.messageSizeThreshold) {
+      return message
+    }
+
+    const payload = JSON.stringify(message)
```

Comment:
Considering that we are dealing with large messages here, solution based on json-stream-stringify would be nicer (we could utilize https://github.com/lokalise/node-core/blob/main/src/utils/streamUtils.ts) here, but this can be done as a separate optimization later

Reply:
Implemented, but I'm not fully convinced that we want to use stream-stringify all the time due to the filesystem overhead. It might be that the final solution should either do everything in memory or allow more granular control over when to dump messages to the filesystem.

Reply:
I agree, it needs to be configureable in some shape. Some users may not be able to use FS at all

Reply:
Made changes as part of https://github.com/kibertoad/message-queue-toolkit/pull/167#discussion_r1643916743. The default behavior is still based on `JsonStreamStringify` + `FsReadableProvider`, but now there is a possibility to override the serialization logic.